[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#88260 [FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`**
* #88235 [FSDP()] Have `fully_shard()` abide by `@contract`!
* #88234 [FSDP()][Easy] Rename `_State` to `_FSDPState`
* #88233 [FSDP()] Rename to `fully_shard()` and move to `_composable/`
* #88232 [FSDP][Easy] Remove unneeded `TrainingState` transition
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87935 [FSDP()][21/N] Refactor and fix `_cast_buffers()`
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward

We can have a separate API for each of the other sharding strategies.

Commit:
[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`

[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]

mrshenli commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -30,7 +30,6 @@
     module: nn.Module,
     process_group: Optional[dist.ProcessGroup] = None,
-    sharding_strategy: Optional[ShardingStrategy] = None,
     mixed_precision: Optional[MixedPrecision] = None,
```
shall we be conservative on adding the args below as well? E.g., can/should replicate/fully_shard share the same mixed_precision API/design? Similar for cpu offload. I assume we can also apply cpu offload for replicate? ignored_modules/param_init_fn/sync_module_states might all need some more discussion.

awgu replied:
Yes, let us discuss the arguments in an upcoming meeting.

I think we need to be careful for `replicate` with CPU offload. I discussed briefly with @rohan-varma, and he mentioned that we want to avoid offloading `world_size`-many copies of the parameters. He had an idea where each rank only offloads `1 / world_size` of the parameters from GPU to some shared CPU memory and each rank reads _all_ parameters from CPU to GPU. I am not sure if that can work in general though.

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`"


We can have a separate API for each of the other sharding strategies.

[ghstack-poisoned]
